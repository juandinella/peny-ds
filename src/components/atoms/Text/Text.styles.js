import { css } from 'styled-components'

import layout from '../../../styles/theme/layout'

export default css`
  line-height: 1.5;

  ${(props) =>
    props.color &&
    css`
      color: var(--color-${(props) => props.color});
    `}

  ${(props) =>
    props.weight &&
    css`
      font-weight: var(--f-${(props) => props.weight});
    `}

  ${(props) =>
    props.size &&
    css`
      font-size: var(
        --font-${(props) => (props.mSize ? props.mSize : props.size)}
      );

      @media screen and (min-width: ${() => layout.breakpoints.sm}) {
        font-size: var(--font-${(props) => props.size});
      }
    `}

  ${(props) =>
    props.lineClamp &&
    css`
      display: -webkit-box;
      line-clamp: ${(props) => props.lineClamp};
      -webkit-line-clamp: ${(props) => props.lineClamp};
      -webkit-box-orient: vertical;
      overflow: hidden;
    `}

    ${(props) =>
    props.breakLine &&
    css`
      word-break: break-all;
    `}

    ${(props) =>
    props.noWrap &&
    css`
      white-space: nowrap;
    `}

  ${(props) =>
    props.breakWord &&
    css`
      word-break: break-all;
    `}

  ${(props) =>
    props.transform &&
    css`
      text-transform: ${(props) => props.transform};
    `}

  ${(props) =>
    props.alignment &&
    css`
      text-align: ${(props) => props.alignment};
    `}
`
