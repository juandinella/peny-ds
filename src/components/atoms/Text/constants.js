export const options = {
  sizes: ['xxs', 'xs', 'sm', 'md', 'lg', 'xl', 'xxl', 'xxxl'],
  weights: ['light', 'regular', 'medium', 'bold'],
  colors: [
    'neutral-120',
    'neutral-100',
    'neutral-80',
    'neutral-60',
    'neutral-40',
    'neutral-20',
    'neutral-00',
    'primary-main',
    'alert-success',
    'alert-error',
  ],
  alignments: ['left', 'center', 'right'],
  types: [
    'div',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'label',
    'legend',
    'p',
    'small',
    'span',
    'strong',
  ],
}
