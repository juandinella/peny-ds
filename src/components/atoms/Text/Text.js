import React from 'react'
import PropTypes from 'prop-types'
import styles from './Text.styles'
import { options } from './constants'
import Skeleton from '../Skeleton'
import Utils from '../../../utils'

/**
 * Display text using well-defined typographic styles.
 */

const Text = ({
  id,
  weight,
  type,
  mSize,
  size,
  color,
  children,
  lineClamp,
  alignment,
  breakWord,
  noWrap,
  breakLine,
  transform,
  shouldRender,
}) => {
  return (
    <Skeleton
      id={`${id}_skeleton`}
      show={!Utils.emptyValue(shouldRender) && !shouldRender()}
    >
      <Text
        css={styles}
        id={id}
        data-testid={id}
        weight={weight}
        size={size}
        mSize={mSize}
        color={color}
        lineClamp={lineClamp}
        alignment={alignment}
        breakWord={breakWord}
        noWrap={noWrap}
        breakLine={breakLine}
        as={type}
        transform={transform}
      >
        {children}
      </Text>
    </Skeleton>
  )
}

Text.displayName = 'Text'

Text.propTypes = {
  /** Text's content */
  children: PropTypes.node.isRequired,
  /** Text's font weight */
  weight: PropTypes.oneOf(options.weights),
  /** Text's font size */
  size: PropTypes.oneOf(options.sizes),
  /** Text's font size for Mobile */
  mSize: PropTypes.oneOf(options.sizes),
  /** Text's color */
  color: PropTypes.oneOf(options.colors),
  /** Text's id */
  id: PropTypes.string,
  /** Text Alignmnet */
  alignment: PropTypes.oneOf(options.alignments),
  /** Don't break the text */
  noWrap: PropTypes.bool,
  /** To break the word */
  breakWord: PropTypes.bool,
  /** If the text is too long it doesn't have any space, this is the way to break it */
  breakLine: PropTypes.bool,
  /** Type of the text */
  type: PropTypes.oneOf(options.types).isRequired,
  /** Text transform short hand. */
  transform: PropTypes.oneOf(options.transforms),
  /** Clamp text */
  lineClamp: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
   * Function that decides if the component should be rendered.
   *
   * @return {boolean} false if component should not be rendered and show skeleton instead. Component renders otherwise.
   */
  shouldRender: PropTypes.func,
}

Text.defaultProps = {
  weight: 'regular',
  size: 'sm',
  color: 'neutral-900',
  type: 'p',
}

export default Text
