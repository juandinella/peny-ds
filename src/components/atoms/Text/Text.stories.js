import Text from './Text'
import { options } from './constants'

import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(Text)
const ListTemplate = getListTemplate(Text)

export default {
  title: 'Atoms/Text',
  component: Text,
  args: {
    children:
      "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
  },
  argTypes: {
    color: getOptionsArgTypes(options.colors),
    size: getOptionsArgTypes(options.sizes),
    mSize: getOptionsArgTypes(options.sizes),
    weight: getOptionsArgTypes(options.weights),
    alignment: getOptionsArgTypes(options.alignments),
    type: getOptionsArgTypes(options.types),
    children: { control: 'text' },
  },
}

export const Default = Template.bind({})

export const Colors = ListTemplate.bind({})
Colors.args = { items: options.colors.map((color) => ({ color })) }

export const Sizes = ListTemplate.bind({})
Sizes.args = { items: options.sizes.map((size) => ({ size })) }

export const Weights = ListTemplate.bind({})
Weights.args = { items: options.weights.map((weight) => ({ weight })) }

export const Skeleton = Template.bind({})
Skeleton.args = { shouldRender: () => false }

export const Alignments = ListTemplate.bind({})
Alignments.args = {
  items: options.alignments.map((alignment) => ({ alignment })),
}
