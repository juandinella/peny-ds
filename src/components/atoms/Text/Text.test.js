import React from 'react'

import { render, cleanup } from '@testing-library/react'

import Text from './Text'

import initStoryshots from '@storybook/addon-storyshots'

initStoryshots({
  storyKindRegex: /^Atoms\/Text$/,
})

const componentProps = {
  id: 'unique_id',
  children: 'Go!',
}

afterEach(cleanup)

describe('<Text />', () => {
  test('should render component correctly', () => {
    const { queryByTestId } = render(<Text {...componentProps} />)

    const p = queryByTestId(componentProps.id)

    expect(p).toBeTruthy()
  })

  test('should render skeleton', () => {
    const { queryByTestId } = render(
      <Text {...componentProps} shouldRender={() => false} />
    )

    expect(queryByTestId(`${componentProps.id}_skeleton`)).toBeTruthy()
  })

  test('should render content correctly', () => {
    const { queryByTestId } = render(<Text {...componentProps} />)

    const p = queryByTestId(componentProps.id)

    expect(p.textContent).toEqual(componentProps.children)
  })

  test('should render type correctly', () => {
    const { container } = render(<Text {...componentProps} type="h4" />)

    const h4 = container.querySelector('h4')

    expect(h4).toBeTruthy()
  })
})
