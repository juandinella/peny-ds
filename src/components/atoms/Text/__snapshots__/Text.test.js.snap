// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Text Alignments 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <p
    className="sc-15mqapg-0 dPGiJx"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 jxGhxN"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 fKWrIq"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
</div>
`;

exports[`Storyshots Atoms/Text Colors 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <p
    className="sc-15mqapg-0 qpaUC"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 kcxBlk"
    color="neutral-100"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 kBUCvR"
    color="neutral-80"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 hMUiWH"
    color="neutral-60"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 gBZAiJ"
    color="neutral-40"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 wBEWf"
    color="neutral-20"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 kBcFHd"
    color="neutral-00"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 jXdSED"
    color="primary-main"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 gASiXl"
    color="alert-success"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 cdnOae"
    color="alert-error"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
</div>
`;

exports[`Storyshots Atoms/Text Default 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <p
    className="sc-15mqapg-0 qpaUC"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
</div>
`;

exports[`Storyshots Atoms/Text Sizes 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <p
    className="sc-15mqapg-0 hjluiu"
    color="neutral-120"
    size="xxs"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 fYWWIm"
    color="neutral-120"
    size="xs"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 qpaUC"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 hMTWpO"
    color="neutral-120"
    size="md"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 iMSdqe"
    color="neutral-120"
    size="lg"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 knzPXW"
    color="neutral-120"
    size="xl"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 hFtcxi"
    color="neutral-120"
    size="xxl"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 kBIYnW"
    color="neutral-120"
    size="xxxl"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
</div>
`;

exports[`Storyshots Atoms/Text Skeleton 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-kjnynt-0 eneJQj"
    data-testid="undefined_skeleton"
    id="undefined_skeleton"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <p
      className="sc-15mqapg-0 qpaUC"
      color="neutral-120"
      size="sm"
    >
      Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
    </p>
  </div>
</div>
`;

exports[`Storyshots Atoms/Text Weights 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <p
    className="sc-15mqapg-0 deqnUi"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 qpaUC"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 ZQvJh"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
  <p
    className="sc-15mqapg-0 iFCGfN"
    color="neutral-120"
    size="sm"
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
  </p>
</div>
`;
